### PR TITLE
[MRG] Add MDD method

### PR DESCRIPTION
**Pull request by ambroiseodt**
Implement the MDD method from https://arxiv.org/pdf/1904.05801. 

Related issue: #254

*Core implementation*
- [x] Add MDDLoss and MDD in ```skada/deep/_adversarial.py```
- [x] Add test_mdd.py in ```skada/deep/tests/test_deep_adversarial.py```
- [x] Update domain_classifier to deal with multiclass in ```skada/deep/modules.py```

*Documentation*
- [x] Update accordingly ```skada/deep/__init__.py```
- [x] Update accordingly ```README.md```
- [x] Update accordingly docs in ```docs/source/all.rst```

### CODE REVIEW

File: skada/deep/_adversarial.py [hunk 1]
```diff
@@ -402,6 +401,151 @@ def CDAN(
...
+        MDD loss. If None, a BCELoss is used.
+
+    References
+    ----------
+    .. [35] Yuchen Zhang et. al. Bridging Theory and Algorithm
+            for Domain Adaptation. In International Conference on
+            Machine Learning, 2019.
+    """
+
+    def __init__(self, gamma=4, domain_criterion=None):
```
Comment (antoinedemathelin) [inline skada/deep/_adversarial.py:429]: gamma=4.

File: skada/deep/_adversarial.py [hunk 2]
```diff
@@ -402,6 +401,151 @@ def CDAN(
...
+        """Compute the domain adaptation loss"""
+        if isinstance(self.domain_criterion_, torch.nn.BCEWithLogitsLoss):
+            pseudo_label_s = y_pred_s < 0
+            pseudo_label_t = y_pred_t < 0
+
+        elif isinstance(self.domain_criterion_, torch.nn.BCELoss):
+            pseudo_label_s = y_pred_s < 0.5
+            pseudo_label_t = y_pred_t < 0.5
+
+        elif isinstance(self.domain_criterion_, torch.nn.BCELoss) or isinstance(
```
Comment (antoinedemathelin) [inline skada/deep/_adversarial.py:456]: Just crossentropy

File: skada/deep/_adversarial.py [hunk 3]
```diff
@@ -402,6 +401,151 @@ def CDAN(
...
+        y_pred_t,
+        domain_pred_s,
+        domain_pred_t,
+        features_s,
+        features_t,
+    ):
+        """Compute the domain adaptation loss"""
+        if isinstance(self.domain_criterion_, torch.nn.BCEWithLogitsLoss):
+            pseudo_label_s = y_pred_s < 0
+            pseudo_label_t = y_pred_t < 0
```
Comment (antoinedemathelin) [inline skada/deep/_adversarial.py:450]: y_preds_s > 0
y_preds_t > 0

File: skada/deep/_adversarial.py [hunk 4]
```diff
@@ -402,6 +401,151 @@ def CDAN(
...
+
+        elif isinstance(self.domain_criterion_, torch.nn.BCELoss):
+            pseudo_label_s = y_pred_s < 0.5
+            pseudo_label_t = y_pred_t < 0.5
```
Comment (antoinedemathelin) [inline skada/deep/_adversarial.py:454]: idem

File: skada/deep/_adversarial.py [hunk 5]
```diff
@@ -402,6 +401,151 @@ def CDAN(
...
+
+def MDD(
+    module,
+    layer_name,
+    reg=1,
+    domain_classifier=None,
+    num_features=None,
+    base_criterion=None,
+    domain_criterion=None,
+    **kwargs,
```
Comment (antoinedemathelin) [inline skada/deep/_adversarial.py:484]: Where is gamma?

File: skada/deep/_adversarial.py (hunk 1 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:429]: :)

File: skada/deep/_adversarial.py (hunk 2 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:456]: :)

File: skada/deep/_adversarial.py (hunk 3 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:450]: :)

File: skada/deep/_adversarial.py (hunk 4 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:454]: :)

File: skada/deep/_adversarial.py (hunk 5 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:484]: :)

File: skada/deep/tests/test_deep_adversarial.py [hunk 6]
```diff
@@ -114,6 +114,55 @@ def test_cdan(domain_classifier, domain_criterion, num_feature, max_feature, n_c
...
+        n_samples_source=n_samples,
+        n_samples_target=n_samples,
+        shift="concept_drift",
+        noise=0.1,
+        random_state=42,
+        return_dataset=True,
+    )
+
+    method = MDD(
+        ToyModule2D(),
```
Comment (YanisLalou) [inline skada/deep/tests/test_deep_adversarial.py:140]: In your case, I assume, it should be `ToyModule2D(n_classes=5, proba=True),` or something like that but clearly not empty otherwise it'll have an output shape of (n, 2)

File: skada/deep/__init__.py [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # Author: Theo Gnassounou <theo.gnassounou@inria.fr>
 #         Remi Flamary <remi.flamary@polytechnique.edu>
 #         Yanis Lalou <yanis.lalou@polytechnique.edu>
-#
+#         Ambroise Odonnat <ambroiseodonnattechnologie@gmail.com>
```
Comment (tgnassou) [inline skada/deep/__init__.py:4]: you removed a line

File: skada/deep/tests/test_deep_adversarial.py [hunk 8]
```diff
@@ -1,16 +1,16 @@
 # Author: Theo Gnassounou <theo.gnassounou@inria.fr>
 #         Oleksii Kachaiev <kachayev@gmail.com>
-#
+#         Ambroise Odonnat <ambroiseodonnattechnologie@gmail.com>
```
Comment (tgnassou) [inline skada/deep/tests/test_deep_adversarial.py:3]: you removed a line

File: skada/deep/_adversarial.py [hunk 9]
```diff
@@ -370,6 +369,148 @@ def CDAN(
...
+    References
+    ----------
+    .. [35] Yuchen Zhang et. al. Bridging Theory and Algorithm
+            for Domain Adaptation. In International Conference on
+            Machine Learning, 2019.
+    """
+
+    def __init__(self, domain_criterion, gamma=4.0):
+        super().__init__()
+        if domain_criterion is None:
```
Comment (tgnassou) [inline skada/deep/_adversarial.py:399]: I know you want to follow the DANN model but it's weird to call them domain_criterion. You can call it disc_criterion or as you want.

File: skada/deep/_adversarial.py [hunk 10]
```diff
@@ -370,6 +369,148 @@ def CDAN(
...
+
+        return disc_loss
+
+
+def MDD(
+    module,
+    layer_name,
+    reg=1,
+    gamma=4.0,
+    domain_classifier=None,
```
Comment (tgnassou) [inline skada/deep/_adversarial.py:435]: same for domain_classifier -> disc_classifier

File: skada/deep/_adversarial.py [hunk 11]
```diff
@@ -370,6 +369,148 @@ def CDAN(
...
+            for Domain Adaptation. In International Conference on
+            Machine Learning, 2019.
+    """
+    if domain_classifier is None:
+        # raise error if num_feature is None
+        if num_features is None:
+            raise ValueError(
+                "If domain_classifier is None, num_features has to be provided"
+            )
+        domain_classifier = DomainClassifier(
```
Comment (tgnassou) [inline skada/deep/_adversarial.py:490]: I don't know if we use here the domain Classifier or if we want a `DiscClassifier` maybe the same that is in the paper

Comment (ambroiseodt) [reply skada/deep/_adversarial.py:490]: In the paper, the disc_classifier must be the same architecture as the task classifier. So in our case, the same than the one used with base_criterion

File: skada/deep/_adversarial.py [hunk 12]
```diff
@@ -370,6 +369,162 @@ def CDAN(
...
+        pseudo_label_s = pseudo_label_s.to(domain_pred_s.device)
+        pseudo_label_t = pseudo_label_t.to(domain_pred_t.device)
+
+        # Deal with CrossEntropyLoss convention in Pytorch
+        if domain_pred_s.shape == pseudo_label_s.shape:
+            domain_pred_s = domain_pred_s.float()
+            pseudo_label_s = pseudo_label_s.float()
+        else:
+            domain_pred_s = domain_pred_s.float()
+            pseudo_label_s = pseudo_label_s.long()
```
[the remark this answers is not in the document]
Comment (ambroiseodt) [reply skada/deep/_adversarial.py]: Not needed anymore

File: skada/deep/tests/test_deep_adversarial.py [hunk 13]
```diff
@@ -114,6 +114,55 @@ def test_cdan(domain_classifier, domain_criterion, num_feature, max_feature, n_c
...
+    "domain_classifier, criterion, num_features",
+    [
+        (DomainClassifier(num_features=10), CrossEntropyLoss(), None),
+        (DomainClassifier(num_features=10), None, None),
+        (None, None, 10),
+    ],
+)
+def test_mdd(domain_classifier, criterion, num_features):
+    module = ToyModule2D()
+    module.eval()
```
[the remark this answers is not in the document]
Comment (ambroiseodt) [reply skada/deep/tests/test_deep_adversarial.py]: Removed. It should be removed for the others too.

File: skada/deep/tests/test_deep_adversarial.py [hunk 14]
```diff
@@ -114,6 +114,55 @@ def test_cdan(domain_classifier, domain_criterion, num_feature, max_feature, n_c
...
+        return_dataset=True,
+    )
+
+    method = MDD(
+        ToyModule2D(),
+        reg=1,
+        gamma=4.0,
+        domain_classifier=domain_classifier,
+        num_features=num_features,
+        criterion=criterion,
```
[the remark this answers is not in the document]
Comment (ambroiseodt) [reply skada/deep/tests/test_deep_adversarial.py]: Done

File: skada/deep/_adversarial.py (hunk 9 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:399]: Done

File: skada/deep/_adversarial.py (hunk 10 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:435]: Done. I also changed domain_pred to disc_pred.

File: skada/deep/_adversarial.py [hunk 15]
```diff
@@ -370,6 +369,150 @@ def CDAN(
...
+    ----------
+    .. [35] Yuchen Zhang et. al. Bridging Theory and Algorithm
+            for Domain Adaptation. In International Conference on
+            Machine Learning, 2019.
+    """
+
+    def __init__(self, disc_criterion, gamma=4.0):
+        super().__init__()
+        if disc_criterion is None:
+            self.disc_criterion_ = torch.nn.CrossEntropyLoss()
```
Comment (antoinedemathelin) [inline skada/deep/_adversarial.py:401]: Is it not possible to set self.disc_criterion_ equal to the task criterion by default?

File: skada/deep/tests/test_deep_adversarial.py [hunk 16]
```diff
@@ -114,6 +115,64 @@ def test_cdan(domain_classifier, domain_criterion, num_feature, max_feature, n_c
...
 
 
+@pytest.mark.parametrize(
+    "disc_classifier, disc_criterion, num_features, n_classes",
+    [
+        (
+            DomainClassifier(num_features=10, n_classes=5),
+            CrossEntropyLoss(),
+            10,
+            5,
```
Comment (antoinedemathelin) [inline skada/deep/tests/test_deep_adversarial.py:125]: Disc_classifier should implement a gradient reversal layer at the beginning to work properly.

**Comment by antoinedemathelin** [comment]
It is almost ok for me
Just two little comments

File: skada/deep/tests/test_deep_adversarial.py (hunk 16 above)
Comment (ambroiseodt) [reply skada/deep/tests/test_deep_adversarial.py:125]: But it is already implemented in DomainClassifier no? What changes do I have to do in the current PR (knowing that we will open another PR to add a DiscClassifier, decorrelate the domain classifier from the modules and so on)?

File: skada/deep/_adversarial.py (hunk 15 above)
Comment (ambroiseodt) [reply skada/deep/_adversarial.py:401]: @antoinedemathelin I can try to set it equal to base_criterion. Would it be ok ?

File: skada/deep/tests/test_deep_adversarial.py (hunk 16 above)
Comment (antoinedemathelin) [reply skada/deep/tests/test_deep_adversarial.py:125]: Yes sorry you have right

File: skada/deep/_adversarial.py (hunk 15 above)
Comment (antoinedemathelin) [reply skada/deep/_adversarial.py:401]: Yes, something like `disc_criterion = copy.deepcopy(base_criterion)` to avoid using the same object for both networks

Comment (ambroiseodt) [reply skada/deep/_adversarial.py:403]: Done

**Comment by antoinedemathelin** [comment]
Sounds good to me!
